python/ad-hoc-ingest/ad_hoc_ingest.py

```diff
@@ -0,0 +1,293 @@
+            metadata["formerRefDept"] = row["formerRefDept"]
+        if row["formerRefTNA"] != "":
+            metadata["formerRefTNA"] = row["formerRefTNA"]
+        if row["checksum_sha256"] == "":
+            metadata["checksum_md5"] = row["checksum_md5"]
+        else:
+            metadata["checksum_sha256"] = row["checksum_sha256"]
+
+        for attempt in range(0,4):
+            try:
+                aws_interactions.upload_file(asset_id, bucket, file_id, get_absolute_file_path(args.input, client_side_path))
+                aws_interactions.upload_metadata(asset_id, bucket, metadata)
+                aws_interactions.send_message(asset_id, bucket, queue_url)
+                break
+            except ClientError as client_error:
+                if attempt == 3:
+                    print(f"Exceeded, number of attempts to recover from error, terminating at file: '{file_id}' from location: '{client_side_path}'")
+                    raise Exception(f"Unable to proceed because: {client_error}. Terminating the process.")
+                else:
+                    print(f"An error caused due to: {client_error}")
+                    input("Fix the error and press any key to continue")
+                    aws_interactions.refresh_session()
+
+        if counter % 10 == 0:
+            print(f"Uploaded ${counter} of ${total}")
+
+# the path in the input file may be relative to the input csv
+def get_absolute_file_path(input_path, relative_or_absolute_file_path):
+    input_file_path = Path(input_path).resolve()
+
+    if Path(relative_or_absolute_file_path).is_absolute():
+        return str(relative_or_absolute_file_path)
+    else:
+        normalised_relative_path = os.path.normpath(relative_or_absolute_file_path.replace("\\", "/"))
+        full_path = Path(input_file_path.parent / normalised_relative_path).resolve()
+        return str(full_path)
+
+def is_folder_writable(output_folder):
+    try:
+        testfile = tempfile.TemporaryFile(dir=output_folder)
+        testfile.close()
+        return True
+    except (OSError, PermissionError):
+        return False
+
+
+def run_ingest(data_set, args, is_upstream_valid):
+    data_set: pandas.DataFrame
+    is_dry_run = False if args.dry_run == False else True
+
+    is_discovery_available = discovery_client.is_discovery_api_reachable()
+    if not is_discovery_available:
+        print("Discovery API is not available for getting metadata information, terminating process")
+        sys.exit(1)
+
+    output_folder = args.output
+    if not is_folder_writable(output_folder):
+        print(f"Unable to write to the output location: '{output_folder}', please make sure that you have necessary permissions for that folder")
+        sys.exit(1)
+
+    prefix = datetime.now().strftime("%Y%m%d_%H%M%S")
+    output_file = os.path.join(output_folder, f"{prefix}_proposed_ingest.csv")
+
+    row_count = 0
+    with open(f"{output_file}", mode="a", newline="", encoding="utf-8") as metadata_csv:
```

Comment:
1. You can change `f"{output_file}"` to just `output_file` since it's a string
2. Maybe this block could go into another method for clarity?